linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+        dataset: &Dataset<ArrayBase<D, Ix2>, T>,
+        mut rng: R,
+    ) -> GaussianMixtureModel<F> {
+        let observations = dataset.records().view();
+        let n_samples = observations.nrows();
+
+        let resp = match hyperparameters.init_method() {
+            GmmInitMethod::KMeans => {
+                let model = KMeans::params_with_rng(hyperparameters.n_clusters(), rng)
+                    .build()
+                    .fit(&dataset);
+                let mut resp = Array::<F, Ix2>::zeros((n_samples, hyperparameters.n_clusters()));
+                for (k, idx) in model.predict(dataset.records()).iter().enumerate() {
+                    resp[[k, *idx]] = F::from(1.).unwrap();
+                }
+                resp
+            }
+            GmmInitMethod::Random => {
+                let mut resp = Array2::<f64>::random_using(
+                    (n_samples, hyperparameters.n_clusters()),
+                    Uniform::new(0., 1.),
+                    &mut rng,
+                );
+                let totals = &resp.sum_axis(Axis(1)).insert_axis(Axis(0));
+                resp = (resp.reversed_axes() / totals).reversed_axes();
+                resp.mapv(|v| F::from(v).unwrap())
+            }
+        };
+
+        let (mut weights, means, covariances) = Self::estimate_gaussian_parameters(
+            &observations,
+            &resp,
+            hyperparameters.covariance_type(),
+            hyperparameters.reg_covariance(),
+        );
+        weights = weights / F::from(n_samples).unwrap();
+
+        // GmmCovarType = full
+        let precisions = Self::compute_precision_cholesky_full(&covariances);
+
+        GaussianMixtureModel {
+            covar_type: *hyperparameters.covariance_type(),
+            weights,
+            means,
+            precisions,
+        }
+    }
+
+    fn estimate_gaussian_parameters<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        _covar_type: &GmmCovarType,
+        reg_covar: F,
+    ) -> (Array1<F>, Array2<F>, Array3<F>) {
+        let nk = resp.sum_axis(Axis(0)) + F::from(10.).unwrap() * F::epsilon();
+        let nk2 = nk.to_owned().insert_axis(Axis(1));
+        let means = resp.t().dot(observations) / nk2;
+        // GmmCovarType = Full
+        let covariances =
+            Self::estimate_gaussian_covariances_full(&observations, resp, &nk, &means, reg_covar);
+        (nk, means, covariances)
+    }
+
+    fn estimate_gaussian_covariances_full<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        nk: &Array1<F>,
+        means: &Array2<F>,
+        reg_covar: F,
+    ) -> Array3<F> {
+        let n_clusters = means.nrows();
+        let n_features = means.ncols();
+        let mut covariances = Array::zeros((n_clusters, n_features, n_features));
+        for k in 0..n_clusters {
+            let diff = observations - &means.slice(s![k..k + 1, ..]);
```

Comment:
means is a 2D NdArray, you can use `Array2::row` here or the generic `Array::index_axis(Axis(0), k))`

Reply:
ok